neo/Ledger/MemoryPool.cs

```diff
@@ -4,162 +4,521 @@
+
+                unsortedPool.Remove(minItem.Transaction.Hash);
+                sortedPool.Remove(minItem);
+            }
+        }
 
-                var exceed = Count - Capacity;
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private bool TryRemoveVerified(UInt256 hash, out PoolItem item)
+        {
+            if (!_unsortedTransactions.TryGetValue(hash, out item))
+                return false;
+
+            _unsortedTransactions.Remove(hash);
+            SortedSet<PoolItem> pool = item.Transaction.IsLowPriority
+                ? _sortedLowPrioTransactions : _sortedHighPrioTransactions;
+            pool.Remove(item);
+            return true;
+        }
 
-                if (exceed > 0)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private bool TryRemoveUnVerified(UInt256 hash, out PoolItem item)
+        {
+            if (!_unverifiedTransactions.TryGetValue(hash, out item))
+                return false;
+
+            _unverifiedTransactions.Remove(hash);
+            SortedSet<PoolItem> pool = item.Transaction.IsLowPriority
+                ? _unverifiedSortedLowPriorityTransactions : _unverifiedSortedHighPriorityTransactions;
+            pool.Remove(item);
+            return true;
+        }
+
+        // Note: this must only be called from a single thread (the Blockchain actor)
+        internal void UpdatePoolForBlockPersisted(Block block, Snapshot snapshot)
+        {
+            _txRwLock.EnterWriteLock();
+            try
+            {
+                // First remove the transactions verified in the block.
+                foreach (Transaction tx in block.Transactions)
                 {
-                    RemoveLowestFee(_mem_pool_free, exceed);
-                    exceed = Count - Capacity;
+                    if (TryRemoveVerified(tx.Hash, out _)) continue;
+                    TryRemoveUnVerified(tx.Hash, out _);
+                }
 
-                    if (exceed > 0)
-                    {
-                        RemoveLowestFee(_mem_pool_fee, exceed);
-                    }
+                // Add all the previously verified transactions back to the unverified transactions
+                foreach (PoolItem item in _sortedHighPrioTransactions)
+                {
+                    if (_unverifiedTransactions.TryAdd(item.Transaction.Hash, item))
+                        _unverifiedSortedHighPriorityTransactions.Add(item);
                 }
+
+                // NOTE: This really shouldn't be necessary any more and can actually be counter-productive, since
+                //       they can be re-added anyway and would incur a potential addtional verification; setting to 1000
+                //       blocks for now. If transactions want to only be valid for short times they can include a script.
+                var lowPriorityCutOffTime = DateTime.UtcNow.AddSeconds(-Blockchain.SecondsPerBlock * 1000);
```

Comment:
In this cutoff new in this PR, @jsolman ?

Reply:
There was a cutoff before, but it was broken because it restamped times of transactions when it rebroadcast after blocks. This fixes it.

Reply:
Hmmm. It is an interesting question. I think it really is good if clients include something to expire transactions explicitly after so many blocks. This cutoff time won't really get the transaction out of the network, because other nodes will still re-relay it after this removes it from this node. So it does seem rather pointless to have it.  Since we can handle the pool full now without issues with these changes it also isn't necessary for stability or anything, so I will go ahead and remove it.